src/chain/chain-training.cc

```diff
@@ -53,21 +57,50 @@ void ComputeChainObjfAndDeriv(const ChainTrainingOptions &opts,
       xent_output_deriv->SetZero();
       numerator.Backward(xent_output_deriv);
     }
+  } else {
+    GenericNumeratorComputation numerator(supervision, nnet_output);
+    num_logprob_weighted = numerator.Forward();
+    KALDI_LOG << "Numerator logprob per frame: "
+              << num_logprob_weighted / (*weight);
+    num_ok = (num_logprob_weighted - num_logprob_weighted == 0);
+    KALDI_LOG << "Numerator Forward " << (num_ok ? "succeeded" : "failed");
```

Comment:
Also remove this. You can probably keep track of how many failed and print somewhere else. But logging all this would be very slow. At least use like KALDI_VLOG(2) or something.

Reply:
I wouldn't worry about speed [it happens only once for each minibatch], but it might be unnecessary.
I could use VLOG.

Reply:
I agree that there should be a count maintained.  If you expect it to normally succeed, maybe you could print something out if it failed?
Also I don't like the name `num_ok`, it's ambiguous.    I prefer `numerator_ok`.

Reply:
It never fails for WSJ. It could fail a few times (for long utterances, only during the first 10 or so iterations) on harder databases (like SWBD). I guess printing something if it fails is enough if you agree? 
I guess maintaining a count involves changing a few different functions/classes (like  ObjectiveFunctionInfo, UpdateStats, and even ComputeChainObjfAndDeriv). Should I do it?

I will fix the name `num_ok`.

Reply:
No, don't change those classes, since they interact with other code.
But there's something else that I'd like you to change, that I just noticed.
Your changes are making the function ComputeChainObjfAndDeriv quite complicated.
I'd like you to, in chain-training.cc but not the header, introduce a new function
ComputeChainObjfAndDerivE2e, and if supervision.e2e is true, call that function.
That will keep the main current code path simple and free of your changes.

Reply:
Will do.